### ISSUE TITLE
Question about yolo_layer.c

### ISSUE THREAD

**Issue by ydixon** [AlexeyAB/darknet#2287]
@AlexeyAB Hi. I've just came from reading https://github.com/AlexeyAB/darknet/issues/1532, https://github.com/AlexeyAB/darknet/issues/2051, https://github.com/AlexeyAB/darknet/issues/1695. I have a few questions.
![delta_yolo_box](https://user-images.githubusercontent.com/22487836/51694750-6ff58a00-203c-11e9-8f75-eaffe39889c0.png)
1. Is it calculating squared error loss between `sigmoid(tx)` instead of `tx` value? Is this intentional?
2. What is `scale` value? Why is it equal to `(2 - truth.w * truth.h)`?

**delta_yolo_class**
https://github.com/AlexeyAB/darknet/blob/6682f0b98984e2b92049e985b21ed81b76666565/src/yolo_layer.c#L142-L145
3. I understand it's using independent binary cross entropy loss for each class. My question is whether the final class loss at the end is:  
- Just sum of the log loss of across all classes or take the average across all classes?
- Let's say if I have a batch of `8`, do we need to divide the total loss by `8`. If true, where can I find it in the code?

Below is some of my interpretation of `forward_yolo_layer`, please let me know if I am wrong.
![yolo_layer_best_n](https://user-images.githubusercontent.com/22487836/51697924-f5c90380-2043-11e9-8592-4919e14454c6.png)

**Comment by ydixon** [comment]
Forgot one question :d
https://github.com/AlexeyAB/darknet/blob/1b15e2f8df5794a8c81fdfafddb3863c9a1a9e16/src/yolo_layer.c#L259-L280

Since the code above is calculating iou for all anchor boxes including the ones that are **not in the current stride layer**, if it's repeated 3 times for yolov3, aren't some of the operations repeated unnecessarily? Is there a reason behind this?

Thanks !

**Comment by AlexeyAB** [comment]
@ydixon Hi,

> Is it calculating squared error loss between sigmoid(tx) instead of tx value? Is this intentional by design?

1. Yes.
     `x[index + 0*stride]` - is sigmoid (logistic_activation), that is calculated here: https://github.com/AlexeyAB/darknet/blob/00b87281f35412dfd0493b879120583951ca279c/src/yolo_layer.c#L432
    * `bbox.x = bx = sigmoid(tx) + cx = x[index + 0*stride] + i`

    * `truth.x = tx`

    * `loss = scale(tx - sigmoid(tx)) = (tx - x[index + 0*stride])`

----

> What is scale value? Why is it equal to (2 - truth.w * truth.h)?

2. Just a bias to get higher delta for small objects, it is required because small objects require more precise positioning and size for higher mAP (to get IoU > 0.5). I don't know why Joseph used exactly this formula.

----

> 3. I understand it's using independent binary cross entropy loss for each class. My question is whether the final class loss calculated at the end is:
> just sum of the log loss across all classes or take the average across all classes?

It takes sum of squares of deltas across all classes.
`loss = delta0*delta0 + delta1*delta1 + ...`

> Let's say if I have a batch of 8, do we need to divide the total loss by 8. If true, where can I find it in the code?

Yes. It is here: https://github.com/AlexeyAB/darknet/blob/00b87281f35412dfd0493b879120583951ca279c/src/network.c#L335

----

> I'm not sure if I understand forward_yolo_layer properly, please let me know if I am wrong.

Yes, we go through all 9 anchors.
Then we get anchor with the best IoU for the current truth-box. (use x=0 y=0 for both)
And we skip anchor if it doesn't correpsond to the anchors whose indexes are indicated in the mask for the current yolo-layer: https://github.com/AlexeyAB/darknet/blob/00b87281f35412dfd0493b879120583951ca279c/cfg/yolov3.cfg#L781

Yes.
Then we calculate loss for: box-coordinates and classes.

----

> Since the code above is calculating iou for all anchor boxes including the ones that are not in the current stride layer, if it's repeated 3 times for yolov3, aren't some of the operations repeated unnecessarily? Is there a reason behind this?

What do you mean?

We shouldn't use truth-box and detected-box at all in the current yolo-layer, if it corresponds to the anchor that isn't specified in `mask=` for the current yolo-layer.
In such a way each yolo layer calculates deltas only for objects with an appropriate size.

**Comment by ydixon** [comment]
@AlexeyAB Really appreciate you spending time to read and answer my question. Thanks!

> 1. Yes.
> `x[index + 0*stride]` - is sigmoid (logistic_activation), that is calculated here:

![paper_tx](https://user-images.githubusercontent.com/22487836/51819580-e381ea80-230d-11e9-8a1e-75959a6c0d9a.png)

I was taking this too literally. Thanks for clarifying. 

> 2. Just a bias to get higher delta for small objects, it is required because small objects require more precise positioning and size for higher mAP (to get IoU > 0.5). I don't know why Joseph used exactly this formula.

This really got me scratching my head. Now I can finally put this to rest :D. I think they did some similar numeric tricks in YoloV2 too.

> It takes sum of squares of deltas across all classes.
> `loss = delta0*delta0 + delta1*delta1 + ...`

Additional question, just to be sure... 
https://github.com/AlexeyAB/darknet/blob/00b87281f35412dfd0493b879120583951ca279c/src/network.c#L335
https://github.com/AlexeyAB/darknet/blob/00b87281f35412dfd0493b879120583951ca279c/src/yolo_layer.c#L304

All these adding/multiplications/squaring operations on the cost doesn't affect the **backpropagation** at all. The only variable that matters is `l.delta`?

> What do you mean?
>
> We shouldn't use truth-box and detected-box at all in the current yolo-layer, if it corresponds to the  anchor that isn't specified in mask= for the current yolo-layer.
> In such a way each yolo layer calculates deltas only for objects with an appropriate size.  

Using YoloV3 as example, there's 3 yololayer. In each yololayer, it's calculating 

1. The IoU between the zero shifted ground truth box `(0,0,w,h)` and the anchor boxes `(0,0,anch_w, anch_h)` for all 9 anchor boxes.
2. Find the `best_iou` and `best_n`
3. Only proceed if `best_n` belongs to this yololayer (determined by `mask_n`)

Given the truth and anchor boxes doesn't change across the 3 yololayers, operation **(1)** and **(2)** seems to be calculating the same values 3 times? Even if the layer size is different for each yololayer, the resulting `best_iou` and `best_n` should be same for all 3 yololayers?

**Comment by AlexeyAB** [comment]
> All these adding/multiplications/squaring operations on the cost doesn't affect the backpropagation at all. The only variable that matters is l.delta?

Yes.


> Given the truth and anchor boxes doesn't change across the 3 yololayers, operation (1) and (2) seems to be calculating the same values 3 times? Even if the layer size is different for each yololayer, the resulting best_iou and best_n should be same for all 3 yololayers?

Yes. It will be calculated 3 times, 1 time for each yolo-layer.

**Comment by ydixon** [comment]
> > All these adding/multiplications/squaring operations on the cost doesn't affect the backpropagation at all. The only variable that matters is l.delta?
> 
> Yes.  

I started trying to find places in code that modify l.delta or does something similar. Then I found this in `update_convolutional_layer_gpu`.
https://github.com/AlexeyAB/darknet/blob/381f90ebb8d1518141e809521fe2985a2d506fc1/src/convolutional_kernels.cu#L846-L851
I can't stop but notice `learning rate / batch`, is this equivalent of dividing `l.delta` by `batch_size`?
  
      


`axpy_ongpu` seems to be pretty common in forward/backward kernel code. But I couldn't really find any information on how to use it and what the parameters do. Would you mind explaining it a little bit? Thanks.

Flow: `axpy_ongpu -> axpy_ongpu_offset -> axpy_kernel`
https://github.com/AlexeyAB/darknet/blob/381f90ebb8d1518141e809521fe2985a2d506fc1/src/blas_kernels.cu#L556-L560
https://github.com/AlexeyAB/darknet/blob/381f90ebb8d1518141e809521fe2985a2d506fc1/src/blas_kernels.cu#L380-L384
